src/chrome/logging/methodsCalledLogger.ts

```diff
@@ -0,0 +1,103 @@
+/*---------------------------------------------------------
+ * Copyright (C) Microsoft Corporation. All rights reserved.
+ *--------------------------------------------------------*/
+import * as _ from 'lodash';
+import { printTopLevelObjectDescription } from './printObjectDescription';
+import { logger } from 'vscode-debugadapter';
+
+enum Synchronicity {
+    Sync,
+    Async
+}
+
+enum Outcome {
+    Succesful,
+    Failure
+}
+
+export class ReplacementInstruction {
+    public constructor(
+        public readonly pattern: RegExp,
+        public readonly replacement: string) { }
+}
+
+export class MethodsCalledLoggerConfiguration {
+    public constructor(readonly replacements: ReplacementInstruction[]) { }
+}
+
+export class MethodsCalledLogger<T extends object> {
+    constructor(
+        private readonly _configuration: MethodsCalledLoggerConfiguration,
+        private readonly _objectToWrap: T,
+        private readonly _objectToWrapName: string) {
+    }
+
+    public wrapped(): T {
+        const handler = {
+            get: <K extends keyof T>(target: T, propertyKey: K, _receiver: any) => {
+                const originalPropertyValue = target[propertyKey];
+                if (typeof originalPropertyValue === 'function') {
+                    return (...args: any) => {
+                        try {
+                            const result = originalPropertyValue.apply(target, args);
+                            if (!result || !result.then) {
+                                this.logCall(propertyKey, Synchronicity.Sync, args, Outcome.Succesful, result);
+                                return result;
+                            } else {
+                                return result.then((promiseResult: unknown) => {
+                                    this.logCall(propertyKey, Synchronicity.Async, args, Outcome.Succesful, promiseResult);
+                                    return promiseResult;
+                                }, (rejection: unknown) => {
+                                    this.logCall(propertyKey, Synchronicity.Async, args, Outcome.Failure, rejection);
+                                    return rejection;
+                                });
+                            }
+                        } catch (exception) {
+                            this.logCall(propertyKey, Synchronicity.Sync, args, Outcome.Failure, exception);
+                            throw exception;
+                        }
+                    };
+                } else {
+                    return originalPropertyValue;
+                }
+            }
+        };
+
+        return new Proxy<T>(this._objectToWrap, handler);
```

Comment:
One possible concern with this is that debugging with Proxies is harder because you don't know what actual object is behind the proxy. Is this going to wrap every object, or how often is it going to be used?

Reply:
At the moment it's wrapping all the objects that the Dependency Injection framework injects.
I'm also planning on adding an option to turn off this logging, so there'll be a way to avoid having the objects being wrapped in the proxies. Does that alleviate your concern?

Reply:
Yeah I think so